src/GovUk.Education.ExploreEducationStatistics.Content.Api.Tests/Controllers/ReleaseFileControllerTests.cs

```diff
@@ -13,40 +15,213 @@
 using GovUk.Education.ExploreEducationStatistics.Content.Api.Tests.Fixtures;
 using GovUk.Education.ExploreEducationStatistics.Content.Model;
 using GovUk.Education.ExploreEducationStatistics.Content.Model.Database;
+using GovUk.Education.ExploreEducationStatistics.Content.Model.Extensions;
+using GovUk.Education.ExploreEducationStatistics.Content.Model.Tests.Fixtures;
+using GovUk.Education.ExploreEducationStatistics.Content.Requests;
 using GovUk.Education.ExploreEducationStatistics.Content.Services.Interfaces;
+using GovUk.Education.ExploreEducationStatistics.Content.ViewModels;
 using Microsoft.AspNetCore.Mvc;
 using Microsoft.AspNetCore.Mvc.Testing;
-using Microsoft.Extensions.DependencyInjection;
 using Moq;
 using Xunit;
 using static GovUk.Education.ExploreEducationStatistics.Common.Services.CollectionUtils;
 using static Moq.MockBehavior;
 
-namespace GovUk.Education.ExploreEducationStatistics.Content.Api.Tests.Controllers
+namespace GovUk.Education.ExploreEducationStatistics.Content.Api.Tests.Controllers;
+
+public abstract class ReleaseFileControllerTests : IntegrationTestFixture
 {
-    public class ReleaseFileControllerTests(TestApplicationFactory testApp) : IntegrationTestFixture(testApp)
+    private ReleaseFileControllerTests(TestApplicationFactory testApp) : base(testApp)
```

Comment:
Do you need this private constructor on the abstract class? I would have thought you could declare the class with a primary constructor like

```csharp
public abstract class ReleaseFileControllerTests(TestApplicationFactory testApp) : IntegrationTestFixture(testApp)
```

Reply:
I originally went with this because I was getting this warning:

> Parameter 'testApp' is captured into the state of the enclosing type and its value is also passed to the base constructor. The value might be captured by the base class as well

However, this was fixable by changing the base `IntegrationTestFixture` constructor to be internal.

This is something that we'll probably run into the more we use primary constructors due to their state capturing nature!

Reply:
From reading up on this it sounds like the captured `testApp` variable can be stored multiple times through different levels of inheritance (especially with our multiple levels of subclasses for partitioning our tests).

If a derived class references `testApp` it will be captured as a separate copy leading to a copy in our test class and a separate version of it in the base class. 

It sounds like the recommended way to go is to avoid using the captured variable and use the protected `TestApp` field from the base class like you've already done.

If we ditch the protected `TestApp` property and made it private wouldn't we still have multiple copies of it?
I see the potential for misuse of `testApp` but maybe it's good enough just to be aware of the warning and not use it? 

I see two places where we should correct it in `Admin.Tests.Services.DataSetVersionServiceTests`.

Reply:
Not sure, would have to test it, but my guess would be that because `testApp` is a reference type this should be fine?

As far as I understand it, this wouldn't result in multiple actual copies of it in memory. It's also less likely to be an issue as the `WebApplicationFactory` APIs are immutable anyway.

It's a shame that there isn't functionality to change the privacy of primary constructor parameters, or a way to pipe them through to the base class. Given the potential issues, it actually feels like the private constructor approach is the safer approach.

Anyway, I'll probably do a little investigation at some point and gather some more evidence / thoughts.